Stream HTML console log

das7pad opened the pull request:
<!-- Comment:
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue ID you created in Jira.
Note that if you want your changes backported into LTS, you need to create a Jira issue. See https://www.jenkins.io/download/lts/#backporting-process for more information.
-->

- Part of https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/704
- Based on https://github.com/jenkinsci/jenkins/pull/11071
- Adopts https://github.com/jenkinsci/stapler/pull/722

This PR is adopting the streaming/unbuffered reads from staplers `LargeText` in `AnnotatedLargeText` and integrates it with the progressive-text in the frontend.

Before: start of request; read all the requested log bytes into memory; emit response with start/end/annotation state in headers
After: start of request; stream the requested log bytes into multi-part response; emit another multi-part part with start/end/annotation state

<!-- Comment:
If the issue is not fully described in Jira, add more information here (justification, pull request links, etc.).

 * We do not require Jira issues for minor improvements.
 * Bug fixes should have a Jira issue to facilitate the backporting process.
 * Major new features should have a Jira issue.
-->

### Testing done

See updated tests for the `log/progressiveHtml` endpoint. Open console while build is in progress.

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Proposed changelog entries

- Stream HTML console log

<!-- Comment:
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/

Do not include the Jira issue in the changelog entry.
Include the Jira issue in the description of the pull request so that the changelog generator can find it and include it in the generated changelog.

You may add multiple changelog entries if applicable by adding a new entry to the list, e.g.
- First changelog entry
- Second changelog entry
-->

### Proposed changelog category

/label rfe

<!--
The changelog entry needs to have a category which is selected based on the label.
If there's no changelog then the label should be `skip-changelog`.

The available categories are:
* bug - Minor bug. Will be listed after features
* developer - Changes which impact plugin developers
* dependencies - Pull requests that update a dependency
* internal - Internal only change, not user facing
* into-lts - Changes that are backported to the LTS baseline
* localization - Updates localization files
* major-bug - Major bug. Will be highlighted on the top of the changelog
* major-rfe - Major enhancement. Will be highlighted on the top
* rfe - Minor enhancement
* regression-fix - Fixes a regression in one of the previous Jenkins releases
* removed - Removes a feature or a public API
* skip-changelog - Should not be shown in the changelog

Non-changelog categories:
* web-ui - Changes in the web UI

Non-changelog categories require a changelog category but should be used if applicable,
comma separate to provide multiple categories in the label command.
-->

### Proposed upgrade guidelines

N/A

<!-- Comment:
Leave the proposed upgrade guidelines in the pull request with the "N/A" value if no upgrade guidelines are needed.
The changelog generator relies on the presence of the upgrade guidelines section as part of its data extraction process.
-->

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

cc @timja @jglick 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

das7pad committed:
Stream HTML console log

Increase buffer sizes in ConsoleAnnotationOutputStream to avoid high
churn of buffers when resetting at each end of a line.

das7pad committed:
Update stapler to pull in interface based constructor for LargeText

das7pad committed:
Add AnnotatedLargeText constructor for interface based LargeText.Source

das7pad committed:
Speed up incremental build

das7pad committed:
Merge remote-tracking branch 'upstream/master' into stapler-upgrade-large-text-interface

das7pad committed:
Run hudson/console/ConsoleAnnotatorTest.java

das7pad committed:
Update stapler to pull in new streaming mode

das7pad committed:
Increase buffer sizes in ConsoleAnnotationOutputStream

This will avoid high churn of buffers when resetting the line buffer at
the end of each line and processing long log lines.

timja committed:
Discard changes to bom/pom.xml

timja committed:
Discard changes to test/pom.xml

timja committed:
Merge branch 'master' into stapler-upgrade-large-text-interface

timja commented on core/src/main/resources/lib/hudson/progressive-text.js:
```diff
@@ -59,9 +60,29 @@ Behaviour.specify(
+          rsp.headers.get("Content-Type")?.startsWith("multipart/form-data")
+        ) {
+          parse = rsp.formData().then((data) => {
+            var text = data.get("text");
+            var meta = JSON.parse(data.get("meta"));
```
```suggestion
            const text = data.get("text");
            const meta = JSON.parse(data.get("meta"));
```

timja commented on test/src/test/java/hudson/console/ConsoleAnnotatorTest.java:
```diff
@@ -147,28 +149,57 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
             WebRequest req = new WebRequest(new URL(r.getURL() + run.getUrl() + "/logText/progressiveHtml" + (start != null ? "?start=" + start : "")));
             req.setEncodingType(null);
-            Map headers = new HashMap();
+            HashMap<String, String> headers = new HashMap<>();
```
```suggestion
            Map<String, String> headers = new HashMap<>();
```

or

```suggestion
            var headers = new HashMap<String, String>();
```

das7pad committed:
Merge remote-tracking branch 'origin/stapler-upgrade-large-text-interface' into stream-html-log

# Conflicts:
#	bom/pom.xml
#	test/pom.xml

das7pad committed:
Address review feedback

- Fix typo in documentation
- use modern JavaScript (let/const/object spread)
- use simpler type for header map
- use descriptive variable in test helper

Co-authored-by: Tim Jacomb <21194782+timja@users.noreply.github.com>

das7pad committed:
Make prettier happy

timja committed:
Update test/src/test/java/hudson/console/ConsoleAnnotatorTest.java

das7pad committed:
Try to fix line ending for windows

timja committed:
Merge branch 'master' into stream-html-log

jglick commented on core/src/main/java/hudson/console/AnnotatedLargeText.java:
```diff
@@ -215,14 +221,13 @@ protected boolean delegateToWriteLogTo(StaplerRequest2 req, StaplerResponse2 rsp
     public long writeLogTo(long start, OutputStream out) throws IOException {
-        return super.writeLogTo(start, new PlainTextConsoleOutputStream(out));
+        if (isHtml())
+            return writeHtmlTo(start, new OutputStreamWriter(out, StandardCharsets.UTF_8));
```
This seems like a potentially incompatible change. Is there a need for it?

das7pad replied:
Indeed, this is not needed. I've built the chain of PRs locally and HTML rendering works in the pipeline-graph-view-plugin. Removed in 2f981dcd02ac7a259f0b5f43c5b19d5bb9764543.

jglick commented on test/pom.xml:
```diff
@@ -52,6 +52,9 @@ THE SOFTWARE.
     <jenkins.javaAgent />
+
+    <!-- Faster incremental build https://github.com/jenkinsci/jenkins/pull/11071#issuecomment-3368149474 -->
+    <test>hudson.console.ConsoleAnnotatorTest</test>
```
:warning:

das7pad replied:
Reverted in aa53098855861b059225c2199fa11d7852f6a41e.

das7pad committed:
Revert unnecessary changes to AnnotatedLargeText.writeLogTo

das7pad committed:
Restore running all tests

das7pad committed:
Merge remote-tracking branch 'upstream/master' into stream-html-log

das7pad committed:
Remove unused import

das7pad committed:
Merge branch 'master' into stream-html-log

das7pad commented:
The stapler upgrade with https://github.com/jenkinsci/stapler/pull/722 is in https://github.com/jenkinsci/jenkins/pull/11221. Once that's merged, I can then remove the incremenal for stapler and mark this PR as ready for review.

das7pad committed:
Merge remote-tracking branch 'upstream/master' into stream-html-log

# Conflicts:
#	bom/pom.xml

jglick reviewed (approved):
Looks OK though I did not test it.

timja reviewed (approved):
Looks reasonable, untested.

-----


/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

jglick commented:
Checked with a simple Pipeline build that printed output at intervals, using built-in console viewer, on Firefox on Chrome. Works, though I am a bit confused since it is still opening a fresh HTTP request every second or so to `…/logText/progressiveHtml` even though it is using chunked output. I thought the point of the new transport was to reuse one HTTP connection for many chunks? Or does that only apply to viewing a _completed_ build (or a sudden very large blob of text)? When I view a completed build, the full text is rendered in the main HTTP response `…/console`. Perhaps I misunderstanding what this PR is supposed to do?

jglick commented:
Ah OK, so the multipart content type here is not used to send multiple chunks of text, merely to work around the fact that HTTP does not permit response headers to be written after content.

A1exKH reviewed (approved):
LGTM.

jglick commented:
(We could really use a PR label on core & plugin PRs forcing a `bom` run. You can do it manually by waiting for Incremental deployment and then filing a draft `weekly-test` PR but it is cumbersome.)
